src/compiler/utilities.rs

```diff
@@ -1,8 +1,58 @@
 #![allow(non_upper_case_globals)]
 
 use std::cmp::Ordering;
+use std::collections::HashMap;
+use std::rc::Rc;
+
+use crate::{
+    create_text_span_from_bounds, insert_sorted, BaseDiagnostic, BaseDiagnosticRelatedInformation,
+    BaseNode, BaseType, Diagnostic, DiagnosticCollection, DiagnosticMessage,
+    DiagnosticRelatedInformationInterface, DiagnosticWithDetachedLocation, DiagnosticWithLocation,
+    Node, NodeInterface, ReadonlyTextRange, SortedArray, SourceFile, SyntaxKind, TextSpan,
+    TypeFlags,
+};
+
+fn get_source_file_of_node<TNode: NodeInterface>(node: &TNode) -> Rc<SourceFile> {
+    if node.kind() == SyntaxKind::SourceFile {
+        unimplemented!()
+    }
+    let mut parent = node.parent();
```

Comment:
This is where we traverse up the AST (via the `.parent` links initialized by the binder) to get the root `SourceFile` given some AST node